src/lib.rs

```diff
@@ -0,0 +1,170 @@
+#[macro_use(matches, _matches_tt_as_expr_hack)]
+extern crate matches;
+extern crate regex;
+use regex::Regex;
+#[macro_use]
+extern crate indoc;
+use std::fs::File;
+use std::io::prelude::*;
+use std::collections::HashMap;
+#[derive(Debug, PartialEq)]
+pub enum IdentType {
+    Space,
+    Tab,
+    None
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Indent {
+    amount: usize,
+    indent: String,
+    ident_type: IdentType
+}
+
+pub fn detect_indent(string: String) -> Indent {
```

Comment:
You'll probably want to make this take an `&str` instead of a `String`. The `String` type is owned, which means if someone passes their string to you like this, they can't get it back. The `&str` type is good for this kind of algorithm, because you're just doing a computation temporarily over the contents of the string but don't need permanent ownership of the string.

The "next level" is to use `AsRef<str>`, as in the examples here:

https://doc.rust-lang.org/1.21.0/std/convert/trait.AsRef.html

That's sort of the most flexible version of the API you can do, even a bit better than `&str`.

Reply:
Thanks. I was actually unsure what was correct here. It wasn't totally obvious what the right approach was. Thanks for the tips.